src/core/graph.js

```diff
@@ -169,80 +169,76 @@ export class Graph<NP, EP> {
   }
 
   /**
-   * Gets the array of all out-edges from the node at the given address.
-   * The order of the resulting array is unspecified.
+   * Find the neighborhood of a given nodeAddress
+   *
+   * By neighborhood, we mean every `{edge, neighbor}` such that edge
+   * has `nodeAddress` as either `src` or `dst`, and `neighbor` is the
+   * address at the other end of the edge.
    */
-  outEdges(
+  neighborhood(
     nodeAddress: Address,
-    typeOptions?: {+nodeType?: string, +edgeType?: string}
-  ): Edge<EP>[] {
+    typeOptions?: {|
+      +nodeType?: string,
+      +edgeType?: string,
+      +direction?: "IN" | "OUT" | "ANY",
+    |}
+  ): {+edge: Edge<EP>, +neighbor: Address}[] {
     if (nodeAddress == null) {
       throw new Error(`address is ${String(nodeAddress)}`);
     }
-    let result = this._lookupEdges(this._outEdges, nodeAddress).map((e) =>
-      this.edge(e)
-    );
 
-    if (typeOptions != null && typeOptions.edgeType != null) {
-      const edgeType = typeOptions.edgeType;
-      result = result.filter((e) => e.address.type === edgeType);
-    }
-    if (typeOptions != null && typeOptions.nodeType != null) {
-      const nodeType = typeOptions.nodeType;
-      result = result.filter((e) => e.dst.type === nodeType);
-    }
-    return result;
-  }
+    let result: {+edge: Edge<EP>, +neighbor: Address}[] = [];
+    const direction = (typeOptions != null && typeOptions.direction) || "ANY";
 
-  /**
-   * Gets the array of all in-edges to the node at the given address.
-   * The order of the resulting array is unspecified.
-   */
-  inEdges(
-    nodeAddress: Address,
-    typeOptions?: {+nodeType?: string, +edgeType?: string}
-  ): Edge<EP>[] {
-    if (nodeAddress == null) {
-      throw new Error(`address is ${String(nodeAddress)}`);
+    if (direction === "ANY" || direction === "IN") {
+      let inNeighbors = this._lookupEdges(this._inEdges, nodeAddress).map(
+        (e) => {
+          const edge = this.edge(e);
+          return {edge, neighbor: edge.src};
+        }
+      );
+
+      if (typeOptions != null && typeOptions.nodeType != null) {
```

Comment:
It should be possible to factor this out to
```js
if (typeOptions != null && typeOptions.nodeType != null) {
  const nodeType = typeOptions.nodeType;
  result = result.filter(({neighbor}) => neighbor.type === nodeType);
}
```
and include the logic just once, above the `edgeType` filter.

Namely, when the following patch is applied, tests still pass:
<details>
<summary>Patch to deduplicate node type filtering logic</summary>

```diff
diff --git a/src/core/graph.js b/src/core/graph.js
index a1f8b89..f90bc31 100644
--- a/src/core/graph.js
+++ b/src/core/graph.js
@@ -197,13 +197,6 @@ export class Graph<NP, EP> {
           return {edge, neighbor: edge.src};
         }
       );
-
-      if (typeOptions != null && typeOptions.nodeType != null) {
-        const nodeType = typeOptions.nodeType;
-        inNeighbors = inNeighbors.filter(
-          ({edge}) => edge.src.type === nodeType
-        );
-      }
       result = result.concat(inNeighbors);
     }
 
@@ -214,14 +207,6 @@ export class Graph<NP, EP> {
           return {edge, neighbor: edge.dst};
         }
       );
-
-      if (typeOptions != null && typeOptions.nodeType != null) {
-        const nodeType = typeOptions.nodeType;
-        outNeighbors = outNeighbors.filter(
-          ({edge}) => edge.dst.type === nodeType
-        );
-      }
-
       if (direction === "ANY") {
         // If direction is ANY, we already counted self-referencing edges as
         // an inNeighbor
@@ -229,9 +214,13 @@ export class Graph<NP, EP> {
           ({edge}) => !deepEqual(edge.src, edge.dst)
         );
       }
-
       result = result.concat(outNeighbors);
     }
+
+    if (typeOptions != null && typeOptions.nodeType != null) {
+      const nodeType = typeOptions.nodeType;
+      result = result.filter(({neighbor}) => neighbor.type === nodeType);
+    }
     if (typeOptions != null && typeOptions.edgeType != null) {
       const edgeType = typeOptions.edgeType;
       result = result.filter(({edge}) => edge.address.type === edgeType);
```
</details>